Rect rotation

dstansby opened the pull request:
## PR Summary
This is a **very** rough draft, but is in a working state if anyone wants to try it. I will add much more context to this PR summary in a bit, but I am opening now just to make people aware I am working on this. **Please don't spend time reviewing this at the moment**. I will take it out of draft status and add the review label when it is ready to be reviewed.

Fixes:
- Allows the rectangle to be rotated past the previous +/- 45 degrees limit
- Fixes scaling the rectangle when it is rotated (to demonstrate, create a selector with rotation and try scaling before and then with this PR)
- Removes the need for `_aspect_ratio_correction`.

## PR Checklist

<!-- Please mark any checkboxes that do not apply to this PR as [N/A]. -->
**Tests and Styling**
- [ ] Has pytest style unit tests (and `pytest` passes).
- [ ] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [ ] New features are documented, with examples if plot related.
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

- A development guide is available at https://matplotlib.org/devdocs/devel/index.html.

- Help with git and github is available at
  https://matplotlib.org/devel/gitwash/development_workflow.html.

- Do not create the PR out of main, but out of a separate branch.

- The PR title should summarize the changes, for example "Raise ValueError on
  non-numeric input to set_xlim".  Avoid non-descriptive titles such as
  "Addresses issue #8576".

- The summary should provide at least 1-2 sentences describing the pull request
  in detail (Why is this change required?  What problem does it solve?) and
  link to any relevant issues.

- If you are contributing fixes to docstrings, please pay attention to
  http://matplotlib.org/devel/documenting_mpl.html#formatting.  In particular,
  note the difference between using single backquotes, double backquotes, and
  asterisks in the markup.

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->

ericpre commented:
This looks very promising!

dstansby commented:
This is ready for review!

jklymak commented:
Confused about the status here - is it waiting, or is it ready?  I'll assume waiting and set to draft, but feel free to pop back on queue.

ericpre commented:
Best to sort out #21977 first before continuing on this PR.

dstansby commented:
This should be good to review now.

ericpre reviewed (commented):
The approach of this PR is much better than the current implementation.
I haven't looked at the code in details yet but I noticed some small issues in the behaviour of the selector:

With the square shape (and without "center" modifier):
![square_modifier_issue](https://user-images.githubusercontent.com/11851990/148431948-02b018d5-01fc-40ea-90c5-33a13733053f.gif)

Rotation issue:
![rotation_issue](https://user-images.githubusercontent.com/11851990/148432084-ab56e1b8-1942-4203-9a35-8fb11d7e5bb3.gif)

Figure resize issue:
![resize_issue](https://user-images.githubusercontent.com/11851990/148433482-23a3aa55-c521-4153-9d96-ea8d76be919a.gif)

dstansby commented:
Thanks for the clear examples - I'll take a look, and try and add a test in each case when fixed.

dstansby commented:
The first two issues were because of negative width and/or height causing confusion with which handle was selected. I've fixed this by ensuring a positive width/height after completing a rectangle, and updated the resize tests to run on 4 different rectangles, each created from a different combination of clicking and dragging from opposite corners.

I've also updated so the tool handles now track the rectangle when the figure is resized. Currently the box tracks the lower left corner of the figure, which perhaps isn't ideal, but as far as I can tell there's no way to keep it fixed in 'absolute full screen coordinates', as a `Figure` does not keep track of its absolute position on screen.

ericpre commented:
There is still an issue with the square modifier (after resetting the selector):
![square_modifier_issue2](https://user-images.githubusercontent.com/11851990/148544866-a1b5db91-d441-4b0d-a23a-e7c3080ef6ac.gif)

I still have the rotation issue as shown in the 2nd example.

Regarding the figure resize issue, the handle are kept in sync with the rectangle patch, but all patches positions are now incorrect. In the [example above](https://github.com/matplotlib/matplotlib/pull/21945#pullrequestreview-845939799), the position of the handle are correct and this is the position of the rectangle patch which is incorrect: the values of the selector shouldn't change as a function of the figure size.

dstansby commented:
Thanks again for testing!

- Square modifier issue is fixed by fixing the setting of _selection_complete when releasing the mouse.
- Rotation issue is fixed by using a fixed value for the center of rotation, instead of recalculating after each mouse move (which results in small errors that add up)

> Regarding the figure resize issue, the handle are kept in sync with the rectangle patch, but all patches positions are now incorrect. In the example above, the position of the handle are correct and this is the position of the rectangle patch which is incorrect: the values of the selector shouldn't change as a function of the figure size.

Because the rectangle is now defined in display coordinates, it's not possible to snap it the handles with non-zero rotation. This is because an scaling the Axes will distort the handles into a parallelogram, not a rectangle.

ericpre commented:
> Because the rectangle is now defined in display coordinates, it's not possible to snap it the handles with non-zero rotation. This is because scaling the Axes will distort the handles into a parallelogram, not a rectangle, so the handles have to snap to the rectangle (in display space).

Considering that the primary aim of the `selectors` is to select range of values, I think that this is a blocking issue: any figure resize will corrupt the selection. We will need to find a workaround or an alternative then

Some minor issues:

Square modifier: the same change is applied to the perpendicular direction, while the width are height should be the same to make the shape square:
![square_modifier_issue3](https://user-images.githubusercontent.com/11851990/148572819-1f2d637a-ec47-469a-92de-f4ed608b33b0.gif)

Rotation issues:
- position incorrect when releasing the mouse button
- click outside the selector reset the rotation
![rotation_issue2](https://user-images.githubusercontent.com/11851990/148572594-a66088e9-2c2b-48c0-87fa-16c43f128f3d.gif)

dstansby commented on lib/matplotlib/tests/test_widgets.py:
```diff
@@ -328,12 +383,13 @@ def onselect(epress, erelease):
                    key=use_key)
-    assert tool.extents == (extents[0] + ydiff, extents[1],
-                            ydata_new, extents[3])
+    assert_allclose(tool.extents, (xdata_new, extents[1],
+                                   extents[2] + xdiff, extents[3]))
```
This the result of fixing a bug in resizing a square selector. Previously:

![Screen Recording 2022-01-08 at 15 12 53](https://user-images.githubusercontent.com/6197628/148649438-4c0e213d-9241-4521-b7a7-31241a3b59fd.gif)

This is now fixed such that when resizing a square selector the edge of the selector always lies on on the cursor.

dstansby commented:
> Considering that the primary aim of the `selectors` is to select range of values, I think that this is a blocking issue: any figure resize will corrupt the selection. We will need to find a workaround or an alternative then

I'm not sure there is a workaround to be had. It's impossible to resize when rotation != 0. We *could* re-scale the selector if rotation =0 and the figure is resized, but I think that would lead to confusion between the rotation != 0 case. Personally I think moving the selection in data coordinates is fine, since the selector is just there to select anything that lies under it in display coordinates, and if you mess with the axes shape/size then the location of your data in display coordinates is going to change anyway.

> Square modifier: the same change is applied to the perpendicular direction, while the width are height should be the same to make the shape square:

Might be fixed, but I'm not clear on exactly what the issue is here, and what the steps are to replicate it?

>     * position incorrect when releasing the mouse button
>     * click outside the selector reset the rotation

Should be fixed

ericpre commented on lib/matplotlib/widgets.py:
```diff
@@ -3267,97 +3338,120 @@ def corners(self):
         Corners of rectangle in data coordinates from lower left,
```
Is this docstring still correct?
When I rotate a square selector by 90 degree, I would get the same values, but it is not the case:
before:
```python
(array([0.34274322, 0.59980032, 0.59980032, 0.34274322]),
 array([0.39714243, 0.39714243, 0.74211084, 0.74211084]))
```
after
```python
(array([0.60062956, 0.5989657 , 0.34191398, 0.34357784]),
 array([0.39826248, 0.74322366, 0.74099079, 0.3960296 ]))
```

ericpre reviewed (commented):
> I'm not sure there is a workaround to be had. It's impossible to resize when rotation != 0. We _could_ re-scale the selector if rotation =0 and the figure is resized, but I think that would lead to confusion between the rotation != 0 case.

If we keep a copy of the position in data coordinates after the selector is modified, either programmatically or interactively, then can apply these values after resizing. This could be done through the `extents` property, for example.

>  Personally I think moving the selection in data coordinates is fine, since the selector is just there to select anything that lies under it in display coordinates, and if you mess with the axes shape/size then the location of your data in display coordinates is going to change anyway.

Other selectors doesn't behave like this and a good example is the [polygon selector example](https://matplotlib.org/stable/gallery/widgets/polygon_selector_demo.html#sphx-glr-gallery-widgets-polygon-selector-demo-py), which keep the same data selection when resizing the figure.
That being said, maybe we could consider using the `use_data_coordinates` argument to define what should be the desired behaviour.

Other than that, it is very good! :)

dstansby commented:
> > I'm not sure there is a workaround to be had. It's impossible to resize when rotation != 0. We _could_ re-scale the selector if rotation =0 and the figure is resized, but I think that would lead to confusion between the rotation != 0 case.
> 
> If we keep a copy of the position in data coordinates after the selector is modified, either programmatically or interactively, then can apply these values after resizing. This could be done through the `extents` property, for example.

We could keep the lower left and the upper right corners fixed in data coordinates? For `rotation != 0` the other two corners would be fixed in data coordinates too, but for `rotation !=0` they would have to change to make sure the corners are still right angles in display coordinates.

dstansby commented on lib/matplotlib/widgets.py:
```diff
@@ -3267,97 +3338,120 @@ def corners(self):
-            self._rotation = np.deg2rad(value)
-            # call extents setter to draw shape and update handles positions
-            self.extents = self.extents
+        self._rotation = np.deg2rad(value)
```
I tried this, but it gets a bit confusing. Because the transform order of a `Rectangle` is

1. Translate the lower left corner to (x0, y0)
2. Rotate

If the rotation is not about (x0, y0), the lower left corner ends up at (x, y) != (x0, y0), which makes for a confusing mess in various bits of other logic.

ericpre commented:
> We could keep the lower left and the upper right corners fixed in data coordinates? For `rotation != 0` the other two corners would be fixed in data coordinates too, but for `rotation !=0` they would have to change to make sure the corners are still right angles in display coordinates.

I am not sure if I am following you and if we are talking about the same, I need to look at the changes in more details. I don't know how it is best to implement it but what I am suggesting to do is the following:
1. after a change of the selector is completed, keep a copy of the selector size, position in data_coordinates
2. in the figure resize callback, apply this copy of the size, position to the selector

In principle, this should be possible to do this with the `extents`. I will look at the changes in more details in the next few days.

ericpre reviewed (commented):
Would it be possible to split commit https://github.com/matplotlib/matplotlib/pull/21945/commits/2e66dde2bff8067ef2181acb80403f71c9491340 into two:
- the first covering the changes from data to display coordinate
- the second covering the changes related to rotation

There are the two main changes of this PR and having them in a single commit is not good for review and the git history.

dstansby commented:
re. resizing the figure canvas, perhaps this is a clearer illustration. The left figure is a rotated selector (on current main branch), and the right figure is after reducing the width of the window. The selector is no longer rectangular, because all four corners stay fixed in data coordinates (and as an aside, it behaves weirdly if you then try to move/rotate it).
<img width="376" alt="Screenshot 2022-01-15 at 19 56 44" src="https://user-images.githubusercontent.com/6197628/149636022-9f4473fe-586d-4db2-8265-87c7353448bf.png"><img width="229" alt="Screenshot 2022-01-15 at 19 56 53" src="https://user-images.githubusercontent.com/6197628/149636024-466c246b-3357-4525-9a4c-6244c3511896.png">

To make sure the selector remains rectangular the two options are:
- Keep one point of the selector (in data or display coords) and width/height in display coordinates fixed
- Keep two points of the selector (e.g. two corners) fixed in data or display coords

I think it makes sense to keep the width/height fixed in display coords, and am undecided on whether a point (either the corner or center?) should be fixed in display or data coords.

> Would it be possible to split commit 2e66dde into two:
>     the first covering the changes from data to display coordinate
>     the second covering the changes related to rotation

These changes are both quite strongly linked, so even if it is possible it would be quite a lot of work to decouple them.

dstansby commented on lib/matplotlib/patches.py:
```diff
@@ -810,6 +810,14 @@ def get_corners(self):
+        Return the edge midpoints of the rectangle, moving anti-clockwise from
+        the center of the left-hand edge.
+        """
+        return self.get_patch_transform().transform(
```
This is private API, and is implicitly tested by a lot of the selector tests. If the points were somehow wrong here, the handles would be in the wrong place, and if they were correct but in the wrong order the resizing logic would not work as intended.

dstansby commented on lib/matplotlib/patches.py:
```diff
@@ -1675,6 +1683,14 @@ def get_corners(self):
+        Return the corners of the ellipse, moving anti-clockwise from
+        the center of the left-hand edge before rotation.
+        """
+        return self.get_patch_transform().transform(
```
See comment above.

dstansby commented on lib/matplotlib/widgets.py:
```diff
@@ -2989,14 +3021,33 @@ def __init__(self, ax, onselect, drawtype='box',
+
+    @property
+    def _position_state(self):
+        """Return a named tuple containing all position state attributes."""
```
I do agree it's not the clearest name, but I think it is descriptive (it returns state that describes the position of the rectangle). Do you have any ideas for an alternative name?

dhomeier commented:
> I think it makes sense to keep the width/height fixed in display coords, and am undecided on whether a point (either the corner or center?) should be fixed in display or data coords.

Looking at it from the data side, so to speak, I would rather give preference to conserving area in data space – holding width and height fixed in data coords would of course be a sufficient, but perhaps not necessary condition for that.
Thinking in particular of the link to astropy/regions#390 here; I don't know if there are other applications in support of this, or if it really should get preference over implementing a better performing solution on the Matplotlib side.

dstansby commented:
> Looking at it from the data side, so to speak, I would rather give preference to conserving area in data space – holding width and height fixed in data coords would of course be a sufficient, but perhaps not necessary condition for that.

What are your thoughts in the instance where the selector is rotated, where it's impossible to keep the selector rectangular (ie. have right angles as corners in display coordinates) **and** conserve both width, height in data coordinates?

dhomeier commented:
> What are your thoughts in the instance where the selector is rotated, where it's impossible to keep the selector rectangular (ie. have right angles as corners in display coordinates) **and** conserve both width, height in data coordinates?

Intuitively it certainly makes interaction easier if angles and aspect ratio are conserved in display coordinates.
Just trying to figure out if the area in data coordinates is not automatically conserved after all – that certainly is the case for 45˚ or 90˚ rotations, as e.g. doubling one side in and aspect ratio of 2 would simply reduce the other by half.
And since the selector is no longer rectangular in data coordinates, the concept of width and height becomes ambiguous anyway – is it to be the parallelogram height or side length?

jklymak commented:
@dstansby have you considered making a simpler "RotatingRectangleSelector" that is mostly meant for the fixed-aspect ratio case?  I really don't see the point of rotating a rectangle if the data dimensions are different in x and y, so this is really mostly for the image processing case?.  If you spin it, it can have whatever whacky behaviour it wants if the data aspect is not 1:1 without introducing confusing behaviour in the normal RectangleSelector?

dstansby commented:
I am not adverse to doing that, but as it currently stands in the main branch (from https://github.com/matplotlib/matplotlib/pull/20839) it is currently possible to rotate a rectangle with axes aspect!=1 and the rectangle stays rectangular in display coordinates, not data coordinates. This PR removes limitations on that approach and fixes bugs with it.

I am happy to either:
- Carry on advocating for this PR
- Revert some of https://github.com/matplotlib/matplotlib/pull/20839 so that `RectangleSelector` is always rectangular in data coordinates (ie. may look like a trapezium in display coordinates), and create a different selector

Either will take significant effort though, so it would be good to know which of these options has support! Personally I think it makes sense to have the rectangle always rectangular in display coords (this PR), but am not against creating a new selector to do this.

jklymak commented:
I added to the weekly meeting this week.  However @dstansby if you would like to attend to discuss we can defer until you are available.  Meetings are Thu 19:00 UTC : https://hackmd.io/49a-u44CTja02xQRaG88JA

jklymak commented:
We discussed on the meeting, and the consensus that the only sane thing is for the rectangle to stay a rectangle in _display_ coordinates.  Anything else doesn't work in any sort of non-cartesian co-ordinate system.  

It also seems that resizing the window, etc is a bit beyond the point.  If you make a selection hopefully something happens with that selection before any other GUI actions are made.  I guess that depends not he use case, but it seems that GUI rectangles can't be expected to respond to window resizing, and anything built on the widget should deal with it another way.

dstansby commented:
thanks for discussing! I'm afraid I don't have the time to dig this up at the moment, as you can probably see it's quite a complex PR!

tacaswell commented:
@dstansby What is the state / fate of this?

dstansby commented:
I think the status is it should definitely go in, but the rebase needs handling, and the what's new and/or bugfixes probably need updating or adding.

I think one of @keflavich or @dhomeier might have some funding from astropy to get this into a mergeable state? Sorry if I'm wrong there, but if I'm right would one of you be happy picking this PR up?

keflavich commented:
There is funding in astropy to do this, but I can't put in the effort.  Someone from Aperio, possibly @dhomeier, may be able to but we haven't worked out details yet.

dhomeier commented:
I am planning to work on this along with astropy/regions#390, but certainly won't be able to start before next year.

timhoffm commented:
Superseded by #26833
